lib/sort.js

```diff
@@ -2,17 +2,18 @@ import moment from 'moment'
 
 import { SID, SIW, SIM, SIY } from './latepenalty'
 
-// Given time t and a promise p, return the absolute distance in seconds
-// between t and the promise's Schelling fence nearest to t.
-// If the deadline is null then return tini minus t, so lack of deadline sorts
-// to the top and, among those, ties are broken so that promises created
-// earlier appear first. (And if promises with no deadlines have creation dates
-// in the future for some reason, those may in fact sort later. That's probably
-// (a) reasonable and (b) almost always moot.)
+// Convert Moment date object to unixtime in seconds
+const unixtm = (d) => { return d.valueOf() / 1000 }
+
+// Given time t and a promise p, return the absolute distance in seconds between
+// t and the promise's Schelling fence nearest to t. If the deadline is null,
+// return a negative number so everything without a deadline sorts to the top,
+// with promises created earlier appearing first. (But if a promise has no
+// deadline and a creation time in the future for some reason then it may in
+// fact sort later. That's probably reasonable and almost always moot.)
 const scheldist = (t, p) => {
-  const tini = p.tini.valueOf() / 1000
-  if (p.tdue === null) { return tini - t }
-  const tdue = p.tdue.valueOf() / 1000
+  if (p.tdue === null) { return p.tini === null ? -1e9 : unixtm(p.tini) - t }
```

Comment:
it might be cleaner for this function to just take two times, and move the null handling to the sort function?

Reply:
The other extreme would also clean this up a bit. We could put all the logic in `scheldist`. If a promise's `tfin` is non-null (it's completed) then return a billion minus `tfin` so it sorts to the end. Subtracting `tfin` means that among the completed promises, more recent ones sort first.

That's like the trick we're already using to get deadline-less promises sorted to the top. So it makes sense to move all this to the same place! The spec is like so... (Recall `tini` is creation date, `tdue` is deadline, and `tfin` is completion date.)

Compute a single metric using a `schelsort` function that takes a time `t` and a promise `p` and returns:

* A huge negative number (say 9e9 = 285 years) if there's something wrong with the promise and it should sort to the very top, like none of `tini`, `tdue`, or `tfin` are set.
* `tini` minus now (typically negative unless `tini` is in the future) if there's no `tdue` or `tfin` set. (If it has no `tdue`, no `tfin`, and `tini` in the future then it could sort later, which is fine, and almost always moot.)
* Absolute distance to nearest Schelling fence for incomplete promises (no `tfin` but `tini` and `tdue` are set).
* A huge positive number, `9e9 - tfin`, if it's completed. Subtracting `tfin` means promises completed later yield a smaller number and sort first.

Then we just sort everything by that single metric!